File: backend/mcp-code-intelligence-python/src/mcp_code_intel/parsers/embedding/chunker.py
"""KSA-169: Chunker — Split text into overlapping chunks for embedding."""
from __future__ import annotations
from dataclasses import dataclass


@dataclass
class Chunk:
    text: str
    index: int
    token_count: int
    start_offset: int
    end_offset: int


class Chunker:
    def __init__(self, max_tokens: int = 512, overlap: int = 128) -> None:
        self._max_tokens = max_tokens
        self._overlap = overlap

    def chunk(self, text: str) -> list[Chunk]:
        words = [w for w in text.split() if w]
        if len(words) <= self._max_tokens:
            return [Chunk(text=text, index=0, token_count=len(words), start_offset=0, end_offset=len(text))]
        chunks: list[Chunk] = []
        start = 0
        chunk_index = 0
        while start < len(words):
            end = min(start + self._max_tokens, len(words))
            chunk_words = words[start:end]
            chunks.append(Chunk(
                text=" ".join(chunk_words), index=chunk_index,
                token_count=len(chunk_words), start_offset=start, end_offset=end,
            ))
            chunk_index += 1
            start += self._max_tokens - self._overlap
            if start >= len(words):
                break
        return chunks

    @property
    def max_tokens(self) -> int:
        return self._max_tokens

    @property
    def overlap(self) -> int:
        return self._overlap

File: backend/src/engine/parsers/embedding/chunker.ts
/**
 * KSA-169: Chunker — Split text into overlapping chunks for embedding.
 * Uses token-based splitting with configurable overlap.
 */

export interface Chunk {
  text: string;
  index: number;
  tokenCount: number;
  startOffset: number;
  endOffset: number;
}

export class Chunker {
  private maxTokens: number;
  private overlap: number;

  constructor(maxTokens: number = 512, overlap: number = 128) {
    this.maxTokens = maxTokens;
    this.overlap = overlap;
  }

  /** Split text into chunks. Returns single chunk if <= maxTokens. */
  chunk(text: string): Chunk[] {
    const words = text.split(/\s+/).filter(Boolean);
    const totalTokens = words.length;

    if (totalTokens <= this.maxTokens) {
      return [{
        text,
        index: 0,
        tokenCount: totalTokens,
        startOffset: 0,
        endOffset: text.length,
      }];
    }

    const chunks: Chunk[] = [];
    let start = 0;
    let chunkIndex = 0;

    while (start < words.length) {
      const end = Math.min(start + this.maxTokens, words.length);
      const chunkWords = words.slice(start, end);
      const chunkText = chunkWords.join(' ');

      chunks.push({
        text: chunkText,
        index: chunkIndex,
        tokenCount: chunkWords.length,
        startOffset: start,
        endOffset: end,
      });

      chunkIndex++;
      start += this.maxTokens - this.overlap;

      if (start >= words.length) break;
    }

    return chunks;
  }

  getMaxTokens(): number {
    return this.maxTokens;
  }

  getOverlap(): number {
    return this.overlap;
  }
}
